### PR TITLE
Conflicts

### PR DESCRIPTION
**Pull request by DavertMik**
This fixes our conflicts functionality.

Conflict rules for partially loaded modules won't be executed.
If module has defined conflict interface, all other modules will be checked for implementing methods of that interface.

In brief:

```
WebDriver+Laravel5 = conflict
WebDriver+PhpBrowser=conflict
Laravel5.ORM + WebDriver = OK
Laravel5.ORM + PhpBrowser = OK
WebDriver + REST (with PHPBrowser) = OK
```

### CODE REVIEW

File: src/Codeception/Lib/ModuleContainer.php [hunk 1]
```diff
@@ -185,18 +177,38 @@ public function injectDependentModule($name, DependsOnModule $module)
     public function validateConflicts()
     {
         $moduleNames = array_keys($this->modules);
-        for ($i = 0; $i < count($this->modules); $i++) {
-            /** @var $currentModule Module  **/
-            $currentModule = $this->modules[$moduleNames[$i]];
+        foreach ($moduleNames as $moduleName) {
+            $currentModule = $this->modules[$moduleName];
+            /** @var $currentModule Module  * */
```
Comment (samdark) [inline src/Codeception/Lib/ModuleContainer.php]: Extra `*` should be removed.

File: src/Codeception/Lib/ModuleContainer.php [hunk 2]
```diff
@@ -185,18 +177,38 @@ public function injectDependentModule($name, DependsOnModule $module)
...
             if (!$currentModule instanceof ConflictsWithModule) {
-                continue;
+                continue; // don't validate modules which are not in conflict
+            }
+            if ($currentModule instanceof PartedModule) {
```
Comment (samdark) [inline src/Codeception/Lib/ModuleContainer.php]: Could be done with `&&`.

File: src/Codeception/Lib/ModuleContainer.php [hunk 3]
```diff
@@ -185,18 +177,38 @@ public function injectDependentModule($name, DependsOnModule $module)
...
-                $inspectedModule = $this->modules[$moduleNames[$j]];
-                $nameAndInterfaces = array_merge([get_class($inspectedModule), $inspectedModule->_getName()], class_implements($inspectedModule));
-                if (in_array(ltrim($currentModule->_conflicts(), '\\'), $nameAndInterfaces)) {
-                    throw new ModuleConflictException($currentModule, $inspectedModule);
+            if (!$this->active[$moduleName]) {
+                continue; // if module is not active it should not be validated
+            }
+
+            $conflicts = $currentModule->_conflicts();
+            if (!interface_exists($conflicts) and !class_exists($conflicts)) {
```
Comment (samdark) [inline src/Codeception/Lib/ModuleContainer.php]: It's better to use `&&` instead of `and`,

File: src/Codeception/Lib/ModuleContainer.php (hunk 2 above)
Comment (DavertMik) [inline src/Codeception/Lib/ModuleContainer.php]: I wanted to show the logic explicitly

Comment (samdark) [inline src/Codeception/Lib/ModuleContainer.php]: Well, that's code style, of course but to me `&&` reads better than nested `if`s.

Comment (DavertMik) [inline src/Codeception/Lib/ModuleContainer.php]: they are not nested, they go one after another

**Comment by DavertMik** [comment]
@janhenkgerritsen @Naktibalda please take a look

**Comment by janhenkgerritsen** [comment]
It looks good to me.

File: tests/data/included/jazz/pianist/tests/functional.suite.yml [hunk 4]
```diff
@@ -8,4 +8,4 @@
 
 class_name: TestGuy
 modules:
-    enabled: [Filesystem, TestHelper]
+    enabled: [Filesystem, \Jazz\Pianist\Codeception\Module\TestHelper]
```
Comment (Naktibalda) [inline tests/data/included/jazz/pianist/tests/functional.suite.yml]: What has this change got to do with conflicts?

**Comment by Naktibalda** [comment]
It looks good to me too.

File: tests/data/included/jazz/pianist/tests/functional.suite.yml (hunk 4 above)
Comment (DavertMik) [inline tests/data/included/jazz/pianist/tests/functional.suite.yml]: Good question! 
https://github.com/Codeception/Codeception/pull/2769/files#diff-366a0aa64f695f98df31e14f2ed8f1e0L65

There is a line that autolocating helpers by namespace is deprecated. I removed it for not to forget to do that in release.

**Pull request merged**